ouroboros-consensus/src-unix/Ouroboros/Storage/IO.hs

```diff
@@ -79,6 +80,15 @@ write :: FHandle -> Ptr Word8 -> Int64 -> IO Word32
 write h data' bytes = withOpenHandle "write" h $ \fd ->
     fromIntegral <$> Posix.fdWriteBuf fd data' (fromIntegral bytes)
 
+-- Like write, but it also makes sure that the whole buffer is flushed.
+writeFull :: FHandle -> Ptr Word8 -> Int64 -> IO ()
+writeFull h data' bytes =
+    let go _ _ 0 = return ()
+        go fd ptr remainingBytes = do
+            writtenBytes <- Posix.fdWriteBuf fd ptr (fromIntegral remainingBytes)
```

Comment:
In contrast to `hGetExactly`, the loop in `writeFull` is not well-founded: if `Posix.fdWriteBuf` starts to return 0 before all bytes have been written, then this will loop forever. I don't know if this can actually happen, but we should at least think about it.